python/lib/core/dmod/core/common/collections/cache.py

```diff
@@ -0,0 +1,1015 @@
+        """
+        Fire the 'on_removal' event
+
+        Args:
+            resolve: Whether to resolve all lingering asynchronous tasks before firing off the event handlers
+        """
+        await self.fire(ON_REMOVAL_KEY, *args, resolve=resolve, **kwargs)
+
+    async def entry_added_async(self, *args, resolve: bool = None, **kwargs):
+        """
+        Fire the 'on_addition' event
+
+        Args:
+            resolve: Whether to resolve all lingering asynchronous tasks before firing off the event handlers
+        """
+        await self.fire(ON_ADDITION_KEY, *args, resolve=resolve, **kwargs)
+
+    async def entry_updated_async(self, *args, resolve: bool = None, **kwargs):
+        """
+        Fire the 'on_update' event
+
+        Args:
+            resolve: Whether to resolve all lingering asynchronous tasks before firing off the event handlers
+        """
+        await self.fire(ON_UPDATE_KEY, *args, resolve=resolve, **kwargs)
+
+    async def entry_accessed_async(self, *args, resolve: bool = None, **kwargs):
+        """
+        Fire the 'on_access' event
+
+        Args:
+            resolve: Whether to resolve all lingering asynchronous tasks before firing off the event handlers
+        """
+        await self.fire(ON_ACCESS_KEY, *args, resolve=resolve, **kwargs)
+
+    def __eq__(self, other):
+        if not isinstance(other, CacheEntry):
+            return False
+
+        return self.__data_hash == other.__data_hash
+
+    def __le__(self, other):
+        if not isinstance(other, CacheEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} <= {other.__class__.__name__}")
+
+        return self.last_accessed <= other.last_accessed
+
+    def __lt__(self, other):
+        if not isinstance(other, CacheEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} < {other.__class__.__name__}")
+
+        return self.last_accessed < other.last_accessed
+
+    def __ge__(self, other):
+        if not isinstance(other, CacheEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} >= {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __gt__(self, other):
+        if not isinstance(other, CacheEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} < {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __str__(self):
+        return f"CacheEntry: {self.identifier}"
+
+    def __repr__(self):
+        return "{" + \
+            f'"identifier": {self.__data}, ' \
+            f'"data_hash": {self.__data_hash}, ' \
+            f'"last_accessed": "{self.last_accessed}"' + \
+            "}"
+
+
+# Part of Issue https://github.com/NOAA-OWP/DMOD/issues/434
+#   "Make metric computations asynchronous by location"
+class AccessCache(typing.Generic[HashableType], typing.MutableMapping[str, CacheEntry[HashableType]]):
```

Comment:
In the same vein as @robertbartel's comment, I think it would be helpful to have a short chat about the design goals of the cache. I too don't see why a `multiprocessing.Queue` isn't sufficient. For large items that are shared between process of thread boundaries `multiprocessing.Value` or `multiprocessing.Array` could be used reduce copying. Views into larger collections could also be exposed as `multiprocessing.Value` or `multiprocessing.Array` proxy objects.

Reply:
I'm slapping together the materials to give a better idea of what this guy is going to do. The code I have stashed that's gonna plug some of this issues of _this_ design utilizes proxies to make everything easier to work with (not to initially write though - that was a nightmare).